piptools/repositories/pypi.py

```diff
@@ -287,6 +287,17 @@ def _get_project(self, ireq):
             return data
         return None
 
+    def _get_download_path(self, ireq):
+        """
+        Determine the download dir location in a way which avoids name
+        collisions.
+        """
+        if ireq.link:
+            salt = hashlib.sha256(ireq.link.url_without_fragment.encode()).hexdigest()
```

Comment:
I've researched pip codebase and have found this [function](https://github.com/pypa/pip/blob/9999f0ec098f9dc4687eecd118f52a3233e64e8c/src/pip/_internal/cache.py#L89-L120). It probably makes sense to do the same for caching files. What do you think?

Reply:
Sure, that looks like it'd work nicely. Only issue would be that it's depending on another pip internal function, so increases the risk of breakage every time they update.
If you're comfortable with that then I can switch it to use that this evening.

Otherwise, I can grab the nesting part of it to give the same behaviour and avoid that dependency.

What's your preference?

Reply:
I would grab the nesting part.

Reply:
OK, that'll be in the next push. Just looking at the test now.

Reply:
Could you update `_get_download_path` in accordance with this [function](https://github.com/pypa/pip/blob/9999f0ec098f9dc4687eecd118f52a3233e64e8c/src/pip/_internal/cache.py#L89-L120)?

Reply:
Can do- but what's the objective?
I can do a straight copy of that function, which will not break on changes to that function's name, but will break if there are changes to the vendored packages that pip uses. Otherwise, what's currently in this PR is a partial copy which should address the concerns but may cause slight cache bloat.
Is the preference here to reduce cache bloat but at the possible expense of stability? If so, I could change it to call that function directly?

Reply:
The intention is to kill two birds with one stone. But, I don't want to block this improvement due to mysterious bugs and rather approve this as is. Thanks for your contribution!